src/canonicity/canonical_chain_discovery.rs

```diff
@@ -205,24 +191,20 @@ pub fn discovery(
 
     let max_canonical_length = canonical_paths
         .last()
-        .and_then(|p| {
-            BlockchainLength::from_path(p)
-                .ok()
-                .map(<BlockchainLength as Into<u32>>::into)
-        })
+        .and_then(|p| extract_block_height(p))
         .unwrap_or(1);
+
     let orphaned: Vec<PathBuf> = paths
         .into_iter()
         .filter(|p| {
-            if let Ok(length) = BlockchainLength::from_path(p) {
-                let length: u32 = length.into();
-                return length <= max_canonical_length && !canonical_paths.contains(p);
+            if let Some(length) = extract_block_height(p) {
+                return length <= max_canonical_length
+                    && !canonical_state_hashes.contains(&extract_state_hash(p));
```

Comment:
Seems to me we could eliminate the block height check and solely rely on the canonical state hashes set. Since orphaned paths are the set difference between all paths and those that are canonical.

Reply:
We're doing the length check because it's not a strict set difference. We have 3 collections:
- `canonical_paths` (up to best block height - canonical threshold)
- `successive_paths` (subtree rooted at the highest block in `canonical_paths`)
- `oprhaned_paths` (the remaining paths)

Reply:
Names still need to be changed